File: mpikurs_solutions/f/p2p_probe_r.f
      program p2p_vector
      implicit none
!  FORTRAN MPI-INCLUDE-file
      include 'mpif.h'
      integer    ierr, nproc, myid,
     :           i, nsend, nrecv, startsend, startrecv
      real*8 vec(10)
!  declare status data structure:
      integer stat(MPI_STATUS_SIZE)

!  initialize MPI, get nproc (# of procs) and myid (local proc. id)
      call MPI_INIT( ierr )
      call MPI_COMM_SIZE( MPI_COMM_WORLD, nproc, ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, myid, ierr )

      do i = 1 , 10
         vec(i) = myid*i 
      end do
      write(6,111)myid,vec
 111  format(i3,'>> initial:    vec=',10f3.0)
           
      if (myid.eq.1) then
         startsend = 5; nsend = 3
         call MPI_SEND( vec(startsend), nsend, MPI_DOUBLE_PRECISION, 0, 
     :                  11, MPI_COMM_WORLD, ierr )
      end if
      
      if (myid.eq.0) then
         startrecv = 2; 
         call MPI_PROBE(MPI_ANY_SOURCE,MPI_ANY_TAG, 
     :                  MPI_COMM_WORLD, stat, ierr)
         call MPI_GET_COUNT(stat, MPI_DOUBLE_PRECISION, nrecv, ierr)
         write(6,*)'number of elements to be received : ',nrecv          
!  status argument "stat" instead of "MPI_STATUS_IGNORE"
         call MPI_RECV(vec(startrecv), nrecv, MPI_DOUBLE_PRECISION, 
     :                 MPI_ANY_SOURCE, MPI_ANY_TAG, 
     :                 MPI_COMM_WORLD, stat, ierr) 
         write(6,113)myid,vec
 113     format(i3,'>> after revc: vec=',10f3.0)
      end if

      call MPI_FINALIZE( ierr )

      stop
      end



 

File: mpikurs_solutions/py/p2p_probe_r.py
#p2p_vector.py
from mpi4py import MPI
import numpy as np

#  create status data structure:
stat = MPI.Status()

comm = MPI.COMM_WORLD
nproc = comm.Get_size()
myid = comm.Get_rank()

vec = myid*np.arange(1,11,dtype = np.float64)
print (myid, '>> initial:  vec=', vec)

if myid == 1:
   nsend = 3
   startsend = 4
   comm.Send(vec[startsend:startsend+nsend],dest=0,tag=0)
if myid == 0:
   comm.Probe(source=MPI.ANY_SOURCE,tag=MPI.ANY_TAG,status=stat)
   nrecv = stat.Get_elements(MPI.DOUBLE)
   print ('number of elements to be received: ',nrecv)
   startrecv = 1
   comm.Recv(vec[startrecv:startrecv+nrecv],source=MPI.ANY_SOURCE,tag=MPI.ANY_TAG)

   print (myid, '>> nach recv: vec=', vec)



